[pallet-revive] EVM backend: Implement tx, block system and call stack instructions

smiasojed opened the pull request:
This PR is part of the road to EVM.
- Implement call and create frames, allowing to call and instantiate other contracts.
- Implement support for tx info, block info, system and contract opcodes.
- The `InstructionResult` <-> `ExecError` conversion functions.

pgherveou committed:
fix

pgherveou committed:
fixes

pgherveou committed:
fixes

pgherveou committed:
fix build

github-actions[bot] committed:
Update from github-actions[bot] running command 'bench --runtime dev --pallet pallet_revive'

pgherveou committed:
Fix seal_call benchmark

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience runtime_dev --bump patch'

github-actions[bot] committed:
Update from github-actions[bot] running command 'bench --runtime dev --pallet pallet_revive'

pgherveou committed:
Merge branch 'master' into pg/fix-seal-call-weights

pgherveou committed:
Merge branch 'pg/fix-seal-call-weights' into pg/eth-decimals

pgherveou committed:
fixes

pgherveou committed:
add migration

pgherveou committed:
get rid of dust_account_id

pgherveou committed:
fixes

pgherveou committed:
lint fix

github-actions[bot] committed:
Update from github-actions[bot] running command 'bench --runtime dev --pallet pallet_revive'

pgherveou committed:
nit updates

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience runtime_dev --bump patch'

pgherveou committed:
fix pallet-xcm test

pgherveou committed:
fix build err

pgherveou committed:
fix assets precompible build

pgherveou committed:
fix

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience runtime_dev --bump major --force'

pgherveou committed:
Merge branch 'master' into pg/eth-decimals

pgherveou committed:
pallet-xcm

pgherveou committed:
Update bench tests

github-actions[bot] committed:
Update from github-actions[bot] running command 'bench --runtime dev --pallet pallet_revive'

pgherveou committed:
Update cargo files

pgherveou committed:
store the runtime code for EVM instantiate

pgherveou committed:
wip

pgherveou committed:
wip

pgherveou committed:
fix

pgherveou committed:
fix

pgherveou committed:
PR review move has_dust and has_balance and remove pub

pgherveou committed:
use local function instead of closure

pgherveou committed:
make migration

pgherveou committed:
scope function

pgherveou committed:
charge 0 for input_data_len on err

pgherveou committed:
comment bench_map

pgherveou committed:
use drain to remove old storage in migration

pgherveou committed:
make balance, dust private in BalanceWithDust

pgherveou committed:
Update substrate/frame/revive/src/migrations/v1.rs

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

pgherveou committed:
Revert "Update substrate/frame/revive/src/migrations/v1.rs"

This reverts commit c94be4fa9b5ab5d37c6a89b5993ef2de7b0569b6.

pgherveou committed:
add migration unit test

pgherveou committed:
typo

pgherveou committed:
Add additional tests

pgherveou committed:
Merge branch 'master' into pg/eth-decimals

pgherveou committed:
Merge branch 'pg/eth-decimals' into pg/revm

pgherveou committed:
update lock

pgherveou committed:
wip

pgherveou committed:
add instructions

pgherveou committed:
wip

pgherveou committed:
split evm / pvm

pgherveou committed:
basic host interaction

pgherveou committed:
fix gas

pgherveou committed:
Merge branch 'master' into pg/revm

pgherveou committed:
wip

pgherveou committed:
update lock

pgherveou committed:
fix

pgherveou committed:
rm

pgherveou committed:
fix

pgherveou committed:
Update

pgherveou committed:
fix

pgherveou committed:
comment bitwise test for now

pgherveou committed:
Add AllowEVMBytecode config

pgherveou committed:
wip

pgherveou committed:
fix compile error

pgherveou committed:
fixes

pgherveou committed:
use 0 value

xermicus committed:
[pallet-revive] revm tests scaffolding (#9290)

- Refactors VM specific tests into dedicated modules.
- Adds the shared VM tests.
- Adds the `solidity-fixtures` crate
- Helper script (`bash fixturs-solidity/build_fixtures.sh`) to
build fixtures.
- Fixtures are checked in for now and to be update manually. Is a KISS
solution. Can be changed to be more complex if it is deemed worth,
  but need to make sure it works for people without Solidity compilers
  and regardless whether the crate is used within the workspace or not.

---------

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>
Signed-off-by: xermicus <cyrill@parity.io>

pgherveou committed:
move back fixture to fixtures folder and generate them from build.rs

disabled AddressPredictor for now

pgherveou committed:
add skeleton for fixtures

pgherveou committed:
wip

pgherveou committed:
fixes

pgherveou committed:
update build.rs

pgherveou committed:
nit

pgherveou committed:
fixes

pgherveou committed:
add missing headers

pgherveou committed:
Merge branch 'master' into pg/revm

pgherveou committed:
add test allow_evm_bytecode_config_works

pgherveou committed:
fix lock

github-actions[bot] committed:
Update from github-actions[bot] running command 'fmt'

pgherveou committed:
fixes

0xRVE committed:
Rve/revm arithmetic instructions WIP (#9361)

add arithmetic instructions

0xRVE committed:
fixes gas computation of evm arithemtic instructions (#9379)

fixes gas computation of evm arithemtic instructions

smiasojed committed:
Add block info instructions to revm

smiasojed committed:
Fix timestamp test

smiasojed committed:
Add tx info instructions

smiasojed commented on substrate/frame/revive/src/vm/evm/instructions/tx_info.rs:
```diff
@@ -18,38 +18,40 @@
-	popn_top!([], index, context.interpreter);
-	let i = as_usize_saturated!(index);
-	*index = context.host.blob_hash(i).unwrap_or_default();
+	gas_legacy!(context.interpreter, revm_gas::BASE);
```
Probably we should charge here a hostFn call base fee only

smiasojed committed:
Add system instructions

smiasojed committed:
Add call instruction support

smiasojed committed:
Improve cross contract calls

smiasojed committed:
Cleanup

smiasojed committed:
Fix error codes

Commit:
Revert "fixes gas computation of evm arithemtic instructions (#9379)"

This reverts commit 5b304a09c42e78da14409efcc4393e9acf6d35e7.

Commit:
Revert "Rve/revm arithmetic instructions WIP (#9361)"

This reverts commit 58b922044c41f5ada449ddae7173569de4e7259e.

xermicus committed:
charge precompile call base costs

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>

xermicus committed:
[pallet-revive] do not silently fail Solidity fixtures compilation (#9474)

This prevents silent failing on compilation errors and using outdated
build artifacts. Instead of ignoring compilation errors and using
outdated fixtures, `cargo test` now bails out early and reports the
problems like this:

```
  Error: failed to compile the Solidity fixtures: [
    {
      "component": "general",
      "errorCode": "2314",
      "formattedMessage": "ParserError: Expected identifier but got '}'\n  --> Flipper.sol:11:1:\n   |\n11 | }\n   | ^\n\n",
      "message": "Expected identifier but got '}'",
      "severity": "error",
      "sourceLocation": {
        "end": 157,
        "file": "Flipper.sol",
        "start": 156
      },
      "type": "ParserError"
    }
  ]
```

Signed-off-by: xermicus <cyrill@parity.io>

xermicus committed:
more generic call tester

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>

xermicus committed:
Merge branch 'pg/revm' into sm/vm-instr

smiasojed committed:
Add CREATE opcode support

pgherveou committed:
refactor more benchmarks

pgherveou commented on substrate/frame/revive/src/exec.rs:
```diff
@@ -1809,6 +1833,45 @@ where
 			.map(|_| address)
 	}
+
+	fn instantiate_with_code(
```
maybe we can merge this with `instantiate` and make the code arg a Code enum to distinguish between hash / bytecode

smiasojed replied:
done

smiasojed committed:
Add Create opcode test

smiasojed committed:
Applied comments

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract.rs:
make sure to replace the context.interpreter.gas calls here as I mentioned in the root PR,
as we should use our own gas meter from extend

pgherveou replied:
add // TODO there as well then

```rust
	let mut gas_limit = context.interpreter.gas.remaining();

	// EIP-150: Gas cost changes for IO-heavy operations
	if context.interpreter.runtime_flag.spec_id().is_enabled_in(SpecId::TANGERINE) {
		// Take remaining gas and deduce l64 part of it.
		gas_limit -= gas_limit / 64
	}
	gas_legacy!(context.interpreter, gas_limit);
```

this has no effect anyway right now since the interpreter.gas is 0

smiasojed commented on substrate/frame/revive/src/vm/evm/instructions/block_info.rs:
```diff
@@ -16,33 +16,36 @@
 /// EIP-1344: ChainID opcode
 pub fn chainid<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-	check!(context.interpreter, ISTANBUL);
 	gas_legacy!(context.interpreter, revm_gas::BASE);
```
Which one is it?

pgherveou replied:
ha sorry you are right there are none

smiasojed committed:
Apply comments

pgherveou commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -91,20 +119,162 @@ pub fn call<'a, E: Ext>(bytecode: Bytecode, ext: &'a mut E, inputs: EVMInputs) -
+
+	// Interpreter call
+	let call_result = interpreter.extend.call(
+		Weight::from_parts(u64::MAX, u64::MAX),
+		U256::MAX,
```
We should add TODO here, eventually the gas model will let us convert the callInputs.gas_limit from gas to weight

For the deposit_limit we should probably also come up with a sane default rather than U256::MAX

smiasojed replied:
I noted this in the PR description under the TODO list: Implement correct Gas handling. For now, my focus was on delivering the functionality needed to run tests on the node with REVM, and I think we’re close to that goal—maybe about a week of work left. The correct gas handling still needs to be properly designed and implemented. Of course I can add TODO comments for both `call` and `create` if you wish.

pgherveou replied:
One thing that is not there as far as I can see is all the logic that lives in pop_frame for pvm
for example that's where we absorb the nested gas_meter and refund the excess of gas that has not been consumed by the frame, not sure if that's what you mean by "implement correct Gas handling"

I would keep TODO marker in the code base to keep track of this for now, but leave that up to you

pgherveou commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -91,20 +119,162 @@ pub fn call<'a, E: Ext>(bytecode: Bytecode, ext: &'a mut E, inputs: EVMInputs) -
+		},
+	};
+
+	let input = match &call_input.input {
+		CallInput::Bytes(bytes) => bytes.to_vec(),
+		// Consider the usage fo SharedMemory as REVM is doing
+		CallInput::SharedBuffer(range) => interpreter.memory.global_slice(range.clone()).to_vec(),
+	};
```
yeah I guess we could refactor call to use &[u8] instead of passing a Vec, I think for now we only ever construct the first variant anyway

xermicus replied:
It looks more like an optimization. Let's fix this later.

pgherveou replied:
yeah and from what I read in the original PR it did not yield great optimisations

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/block_info.rs:
```diff
@@ -16,33 +16,36 @@
 ///
 /// Pushes the current block's beneficiary address onto the stack.
 pub fn coinbase<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 	gas_legacy!(context.interpreter, revm_gas::BASE);
```
RuntimeCosts::BlockAuthor

smiasojed committed:
Create2 works

smiasojed committed:
Add delegate and call instructions

pgherveou committed:
Merge branch 'master' into pg/revm

pgherveou committed:
Merge branch 'master' into pg/revm

pgherveou committed:
revm file shuffling

pgherveou committed:
fixes

pgherveou committed:
rm stuff for later

pgherveou committed:
rm evm

pgherveou committed:
rm evm stuff

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience runtime_dev --bump patch'

pgherveou committed:
rm

pgherveou committed:
taplo fix

pgherveou committed:
fix up cargo.lock

pgherveou committed:
try

pgherveou committed:
fixes

pgherveou committed:
Merge branch 'master' into pg/revm-base

pgherveou committed:
rm vm/runtime.rs

pgherveou committed:
Merge branch 'pg/revm-base' into pg/revm

pgherveou committed:
fix

pgherveou committed:
nit

smiasojed committed:
Add gas and  keccak256 instructions

pgherveou committed:
fixes

smiasojed committed:
Address comments

pgherveou committed:
nit

pgherveou committed:
update call_with_code_per_byte

pgherveou committed:
fixes

pgherveou committed:
added migration

pgherveou committed:
Merge branch 'master' into pg/revm-base

pgherveou committed:
Merge branch 'pg/revm-base' into pg/revm

pgherveou committed:
refactoring

pgherveou committed:
fix clippy

smiasojed committed:
Merge remote-tracking branch 'origin/pg/revm' into sm/vm-instr

pgherveou committed:
fix clippy

smiasojed committed:
Add instantiate with code type support

smiasojed committed:
Add system instruction tests

pgherveou committed:
Merge branch 'master' into pg/revm

pgherveou committed:
merge fix

pgherveou committed:
simplify migration tests

pgherveou committed:
add migration

pgherveou committed:
remove unneeded tests for this PR

pgherveou committed:
fixes

pgherveou committed:
keep these files for next PR

pgherveou committed:
comments

pgherveou committed:
fixes

pgherveou committed:
fix

github-actions[bot] committed:
Update from github-actions[bot] running command 'bench --runtime dev --pallet pallet_revive'

smiasojed committed:
Add system instr and tests cleanup

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience runtime_dev --bump patch'

pgherveou committed:
rename seal_to_account_id

pgherveou committed:
update PRDOC

pgherveou committed:
Merge branch 'master' into pg/revm

pgherveou committed:
fix bench

smiasojed committed:
Improve gas handling for CREATE intruction

pgherveou committed:
output.data shoudl be untouched

pgherveou committed:
install solidity in job

pgherveou committed:
use resolc too

pgherveou committed:
rm line

pgherveou committed:
enforce evm code size limit

pgherveou committed:
fix

pgherveou committed:
fix

pgherveou committed:
fix

pgherveou committed:
PR review

smiasojed committed:
Code cleanup and calls refactoring

xermicus commented on substrate/frame/revive/src/exec.rs:
```diff
@@ -265,11 +264,26 @@ pub trait PrecompileWithInfoExt: PrecompileExt {
 		salt: Option<&[u8; 32]>,
 	) -> Result<H160, ExecError>;
+
+	// /// Instantiate a contract from the given bytecode.
+	// ///
+	// /// Returns the original code size of the called contract.
+	// /// The newly created account will be associated with `code`. `value` specifies the amount of
+	// /// value transferred from the caller to the newly created account.
+	// fn instantiate_with_code(
+	// 	&mut self,
+	// 	gas_limit: Weight,
+	// 	deposit_limit: U256,
+	// 	code: Vec<u8>,
+	// 	value: U256,
+	// 	input_data: Vec<u8>,
+	// 	salt: Option<&[u8; 32]>,
+	// ) -> Result<H160, ExecError>;
```
```suggestion
```

smiasojed committed:
Update substrate/frame/revive/src/exec.rs

Co-authored-by: xermicus <cyrill@parity.io>

pgherveou committed:
make pristine unbounded and check PVM & EVM code size

xermicus committed:
create: set caller frame

Signed-off-by: xermicus <cyrill@parity.io>

pgherveou committed:
nit

pgherveou committed:
refcount = 1 for evm

xermicus committed:
the exec error to halt reason mapping function

Signed-off-by: xermicus <cyrill@parity.io>

pgherveou committed:
fix

pgherveou committed:
fix

pgherveou committed:
add one more assert

pgherveou committed:
nit

xermicus committed:
the callcode instruction is unsupported

Signed-off-by: xermicus <cyrill@parity.io>

xermicus committed:
typos

Signed-off-by: xermicus <cyrill@parity.io>

xermicus committed:
typos

Signed-off-by: xermicus <cyrill@parity.io>

xermicus committed:
format the fixtures

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>

xermicus reviewed (commented):
I added the exec error to halt reason mapping function but some logic need to be changed. The current implementation has some divergences in error handling, for example `create2` halts when the target address already exist or the callee went out of gas (those are recoverable errors and it should be implemented according to the EVM spec). All failure cases should be tested too (the test coverage of call and create family opcodes is too sparse). They are well documented on [evm.codes](evm.codes).

pgherveou committed:
Remove refcount and owner for EVM CodeInfo

pgherveou committed:
rm unused

smiasojed commented:
> ## Implement correct Gas handling
> ## for memory instructions
> We will use PVM memory limit instead of charging for memory expension
> 
> ## for contracts instruction, copy from/to contract
> which instructions exactly ? Maybe we can add new benchmarks where it make sense.

We already have benchmarks for copy operations for PVM.
For REVM integration, we should probably benchmark:
RuntimeCosts::CopyToContract(len) → input calldata copied into contract memory.
RuntimeCosts::CopyFromContract(len) → returndata copied back to the caller.
Both of these are exercised implicitly during CALL / DELEGATECALL / CREATE execution.

In system opcodes the `memory_resize` function in REVM already accounts for memory expansion costs.
It also charges copy costs. We should consider incorporating our CopyFromContract logic into this process to align with pallet-revive costs.

Call Base Cost
Currently, base call costs (RuntimeCosts::CallBase, DelegateCallBase, etc.) are benchmarked using PVM.
We should evaluate whether to also introduce benchmarks for EVM (REVM) calls to ensure consistency across both implementations.

> 
> ## how to handle gas limits, deposit limits - now we use MAX
> this might requires Alex's gas model PR to land first
> 

Agree

> ## Implement benchmarks for custom REVM flows
> ## benchmark chain id
> we can keep gas_legacy!(context.interpreter, revm_gas::BASE); here since it's just basically returning a constant or better benchmark the function in PVM and apply the same benchmark
> 
> ## benchmark instantiate with code
> should be similar to existing benchmark
> 

yes, it is

> ## benchmark system instructions
> we can keep gas_legacy!
> 

we want to keep copy cost as they are currently handled in REVM?

> ## Implement stack/interpreter tests, execute calls between PVM/EVM
> should be easy to add some tests here
> 

yes, it is easy byt will take 2-3 days 
Implement cross-VM call scenarios to test interoperability:
- EVM → EVM
- PVM → EVM
- EVM → PVM

Check that input calldata and output returndata are copied correctly.
Test revert propagation across frames:
Recoverable errors (RETURN, REVERT) vs unrecoverable (OUT_OF_GAS, invalid opcodes).
Ensure InstructionResult::Revert is handled consistently.
Confirm that nested calls correctly propagate revert data back to the original caller.

> ## Add tracing when running in REVM interpreter
> call / prestate tracing should work already, we can add opcode tracing later if needed
> 
Agree
> ## Add system instruction tests: gasleft, codecopy
> should not take too long
> 
> ## Check how to handle warm/cold storage
> We treat everything as cold for now
> 
Agree
> ## Implement missing CALLCODE contract instructions
> We can turn it off for now and keep it as a todo for later, or move it to a new PR
> 
> ## Implement conversion between REVM Interpreter and pallet revive errors
> Not sure what you mean here, but does not sound like a huge task
>

The error handling/revert propagation is currently incomplete; at the moment, all errors are treated as out-of-gas.

> ## Check data sharing using shared memory between frames
> there should be no data shared between frames apart from the returned data
> 
> ## Check/Implement correctly EIP-3541, EIP-170, EIP-7907, EIP-7702, EIP-158, EIP-2 point 3, EIP-3607
> Aren't these already implemented in REVM. We should consider all these enabled, EIP-7702 tx will be rejected for now
> 
There are checks in REVM like:
```
   // EIP-3541: Reject new contract code starting with the 0xEF byte
    if !is_eip3541_disabled
        && spec_id.is_enabled_in(LONDON)
        && interpreter_result.output.first() == Some(&0xEF)
    {
        journal.checkpoint_revert(checkpoint);
        interpreter_result.result = InstructionResult::CreateContractStartingWithEF;
        return;
    }
```
It is part of integration with REVM frame so it is not present now in our codebase.

> ## Compare REVM memory consumption with PVM to ensure it fits the memory budget
> We use the same memory budget for both, I guess we just need to ensure that it's safe
> 
I think we need to:
Set the appropriate memory limit for REVM.
Measure and analyze REVM memory usage during typical execution.
Identify any internal caches or persistent allocations that might need clearing.
Document differences versus PVM memory behavior and adjust configuration if necessary.

> @smiasojed addressed the TODO, cc @xermicus @athei can you review / comment, so we can figure out a rough timeline estimation

pgherveou committed:
origin should be held not caller

pgherveou committed:
Merge branch 'master' into pg/revm

smiasojed committed:
Rm copy cost from call

pgherveou committed:
rollback refcount removals

pgherveou committed:
Update Cargo.toml

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

pgherveou committed:
Update prdoc/pr_9285.prdoc

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

pgherveou committed:
Update substrate/frame/revive/src/call_builder.rs

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

pgherveou committed:
Update substrate/frame/revive/src/benchmarking.rs

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

pgherveou committed:
rm charge_evm_init_cost

pgherveou committed:
fix

pgherveou committed:
rm - 1

pgherveou committed:
restore legacy behaviour

smiasojed committed:
Add revert test and fix errors mapping

smiasojed committed:
Add call tests

xermicus committed:
the unified error mapping approach

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>

smiasojed committed:
Enable invalid opcode test for EVM

pgherveou committed:
update

xermicus committed:
the instruction result to exec error mapping function

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>

pgherveou committed:
deadcode unused for now

xermicus committed:
the INVALID EVM opcode test

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>

smiasojed committed:
Reject bytecode starting with 0xef (EOF)

xermicus committed:
Merge branch 'pg/revm' into sm/vm-instr

xermicus committed:
the create tester needs value to instantiate contracts

Signed-off-by: Cyrill Leutwiler <bigcyrill@hotmail.com>

pgherveou committed:
deposit fixes

- CodeUploadDeposit is transferred from a pallet-account instead of the
  origin
- Migration takes care of moving the hold from the owner to the pallet's
  account
- Remove code extrinsic is gone, it happens automatically when refcount
  drops to 0

pgherveou committed:
add missing assert

pgherveou committed:
tweak tests

pgherveou committed:
ensure pallet_account exits

smiasojed committed:
Use U256 converter

pgherveou committed:
fixes

pgherveou committed:
nit

pgherveou committed:
update comment

pgherveou committed:
Add back remove_code

smiasojed committed:
Merge remote-tracking branch 'origin/pg/revm' into sm/vm-instr

smiasojed committed:
Add missing license

pgherveou committed:
nit

pgherveou commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -15,27 +15,55 @@
+/// For `pallet-revive`, this is hardcoded to 0
+pub(crate) const BASE_FEE: U256 = U256::zero();
+
+/// EVM max deployed runtime code size (EIP-170).
+/// 24,576 bytes (0x6000).
+pub const EVM_BYTECODE_LIMIT: usize = 24_576;
+
+/// EVM max initcode size (EIP-3860).
+/// 49,152 bytes (0xC000).
+pub const EVM_INITCODE_LIMIT: usize = EVM_BYTECODE_LIMIT * 2;
```
these constants are defined already by 
revm::primitives::eip3860::MAX_INITCODE_SIZE 
and 
revm::primitives::eip170::MAX_CODE_SIZE 

let's not redefine them, these const are used below in from_evm_runtime / from_init_runtime

smiasojed replied:
done

pgherveou commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -44,7 +72,17 @@ impl<T: Config> ContractBlob<T> {
+		}
+
 		let code_len = code.len() as u32;
+		// EIP-7907: Code limit set to 0xc000 (~49kb).
+		if code_len as usize > EVM_BYTECODE_LIMIT {
+			return Err(<Error<T>>::BlobTooLarge.into());
+		}
```
this is checked above already

smiasojed replied:
done

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract/call_helpers.rs:
```diff
@@ -64,25 +68,48 @@ pub fn resize_memory<'a, E: Ext>(
 	Some(offset..offset + len)
 }
 
-/// Calculates gas cost and limit for call instructions.
```
let's keep some doc str

smiasojed replied:
done

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract/call_helpers.rs:
```diff
@@ -64,25 +68,48 @@ pub fn resize_memory<'a, E: Ext>(
 
-	Some(gas_limit)
+			// Cost for decoding input
+			gas!(interpreter, RuntimeCosts::PrecompileDecode(input_len as u32), None);
```
why don't we charge for RuntimeCosts::CopyFromContract(input_len) for the None branch?

smiasojed replied:
Added

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract.rs:
```diff
@@ -50,16 +56,27 @@ pub fn create<'ext, const IS_CREATE2: bool, E: Ext>(context: Context<'_, 'ext, E
 	popn!([value, code_offset, len], context.interpreter);
 	let len = as_usize_or_fail!(context.interpreter, len);
 
+	// TODO: We do not charge for the new code in storage. When implementing the new gas:
```
can we link to the issue if we have it already

smiasojed replied:
done

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract.rs:
```diff
@@ -71,33 +88,22 @@ pub fn create<'ext, const IS_CREATE2: bool, E: Ext>(context: Context<'_, 'ext, E
 	let scheme = if IS_CREATE2 {
 		popn!([salt], context.interpreter);
-		// SAFETY: `len` is reasonable in size as gas for it is already deducted.
-		gas_or_fail!(context.interpreter, revm_gas::create2_cost(len));
```
you don't charge at all here?

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract.rs:
```diff
@@ -121,36 +127,26 @@ pub fn call<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
+	// TODO: Handle the cold/warm storage correctly
+	let scheme = CallScheme::Call;
+	let input = CallInput::SharedBuffer(input);
```
We won't that will probably be baked in into frame if we do by checking whether or not the accessed storage is in the cache

nit: let's inline this as in the original code

smiasojed replied:
you mean to inline scheme and input? I am using it now also in calc_call_gas not only in action

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract.rs:
```diff
@@ -216,38 +170,34 @@ pub fn delegate_call<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 	let to = Address::from_word(B256::from(to));
 	// Max gas limit is not possible in real ethereum situation.
-	let local_gas_limit = u64::try_from(local_gas_limit).unwrap_or(u64::MAX);
+	let _local_gas_limit = u64::try_from(local_gas_limit).unwrap_or(u64::MAX);
```
let's mention the issue where this should be fixed if we can

smiasojed replied:
done

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract.rs:
```diff
@@ -216,38 +170,34 @@ pub fn delegate_call<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
+	// TODO: Handle the cold/warm storage correctly
+	let scheme = CallScheme::DelegateCall;
+	let input = CallInput::SharedBuffer(input);
```
nit let's inline

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/contract.rs:
```diff
@@ -261,36 +211,34 @@ pub fn static_call<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-	// Set `is_empty` to false as we are not creating this account.
-	load.is_empty = false;
-	let Some(gas_limit) = calc_call_gas(context.interpreter, load, false, local_gas_limit) else {
+	// TODO: Handle the cold/warm storage correctly
+	let scheme = CallScheme::StaticCall;
+	let input = CallInput::SharedBuffer(input);
```
let's inline

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/system.rs:
```diff
@@ -213,8 +230,9 @@ pub fn returndatacopy<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 pub fn gas<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-	gas_legacy!(context.interpreter, revm_gas::BASE);
-	push!(context.interpreter, U256::from(context.interpreter.gas.remaining()));
+	gas!(context.interpreter, RuntimeCosts::RefTimeLeft);
+	let gas = context.interpreter.extend.gas_meter().gas_left().ref_time();
```
maybe mention a TODO here

smiasojed replied:
done

pgherveou commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -110,32 +148,148 @@ pub fn call<'a, E: Ext>(bytecode: Bytecode, ext: &'a mut E, inputs: EVMInputs) -
-fn run<WIRE: InterpreterTypes>(
-	interpreter: &mut Interpreter<WIRE>,
-	table: &revm::interpreter::InstructionTable<WIRE, DummyHost>,
+/// Runs the EVM interpreter until it returns an action.
```
```suggestion
/// Runs the EVM interpreter
```

or and handle actions if you prefer

pgherveou commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -110,32 +148,148 @@ pub fn call<'a, E: Ext>(bytecode: Bytecode, ext: &'a mut E, inputs: EVMInputs) -
+		let action = interpreter.run_plain(table, host);
+		match action {
+			InterpreterAction::Return(result) => {
+				log::info!("Return {:?}", result);
```
use debug and LOG_TARGET

pgherveou replied:
can you add that as well

```
			debug_assert!(
				result.gas.limit() == 0 &&
					result.gas.remaining() == 0 &&
					result.gas.refunded() == 0,
				"Interpreter gas state should remain unchanged; found: {:?}",
				result.gas,
			);
```

to make sure we don't use the gas field here

smiasojed replied:
why do we need this assert?

pgherveou replied:
want to make sure we haven't left any instructions that touches the result.gas that we don't use

smiasojed committed:
Address comment

smiasojed committed:
Update substrate/frame/revive/src/vm/evm.rs

Co-authored-by: PG Herveou <pgherveou@gmail.com>

pgherveou committed:
Update substrate/frame/revive/src/lib.rs

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

pgherveou committed:
mint balance in migration

pgherveou committed:
rm check

pgherveou committed:
fmt

pgherveou committed:
fixes

pgherveou committed:
reject hash == 0

pgherveou committed:
warm up pallet_account

pgherveou committed:
Update substrate/frame/revive/src/migrations/v2.rs

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

pgherveou committed:
Update ExtBuilder

smiasojed committed:
Apply comments

smiasojed commented on substrate/frame/revive/fixtures/contracts/Caller.sol:
```diff
@@ -0,0 +1,55 @@
+            addr := create2(0, add(initcode, 0x20), mload(initcode), salt)
+            if iszero(addr) {
+                // bubble failure
+                revert(0, 0)
```
Is it still valid and shall be address in this PR?

smiasojed committed:
Fmt

smiasojed committed:
Add assert for gas checking

pgherveou committed:
Merge branch 'master' into pg/revm

pgherveou committed:
Update .github/workflows/tests-misc.yml

github-actions[bot] committed:
Update from github-actions[bot] running command 'bench --runtime dev --pallet pallet_revive'

pgherveou committed:
fix tests-misc

pgherveou committed:
fix

pgherveou committed:
Merge branch 'pg/revm' into sm/vm-instr

pgherveou committed:
Merge branch 'master' into sm/vm-instr

pgherveou commented on substrate/frame/revive/src/vm/evm/instructions/tx_info.rs:
```diff
@@ -15,41 +15,41 @@
 /// Implements the GASPRICE instruction.
 ///
 /// Gets the gas price of the originating transaction.
 pub fn gasprice<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
```
needs some update as well once the new gas model is merged

xermicus replied:
Yeah - do you want me to add a `TODO`? FWIW think we will have a plethora of `TODO`s anyways so not sure if it's worth to add them everywhere. Will have to go over everything again anyways.

pgherveou committed:
fix log level

pgherveou commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -111,32 +136,157 @@ pub fn call<'a, E: Ext>(bytecode: Bytecode, ext: &'a mut E, inputs: EVMInputs) -
+
+	let salt = match create_input.scheme {
+		CreateScheme::Create => None,
+		CreateScheme::Create2 { salt } => {
+			let mut arr = [0u8; 32];
+			arr.copy_from_slice(salt.as_le_slice());
+			Some(arr)
+		},
```
```suggestion
		CreateScheme::Create2 { salt } => Some(salt.to_le_bytes()),
```

pgherveou committed:
fix

xermicus committed:
the caller fixture bubbles up errors

Signed-off-by: xermicus <cyrill@parity.io>

xermicus committed:
add missing instruction error handling

Signed-off-by: xermicus <cyrill@parity.io>

xermicus commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -197,13 +347,86 @@ impl InputsTr for EVMInputs {
+		InstructionResult::OverflowPayment => Some(Error::<E::T>::BalanceConversionFailed),
+		InstructionResult::CreateContractSizeLimit | InstructionResult::CreateInitCodeSizeLimit =>
+			Some(Error::<E::T>::StaticMemoryTooLarge),
+		_ => None,
```
Good catch - changed it into an exhaustive `match`, found some not covered yet 😅

xermicus committed:
Update substrate/frame/revive/src/exec.rs

Co-authored-by: Alexander Theißen <alex.theissen@me.com>

xermicus committed:
rename the trait method

Signed-off-by: xermicus <cyrill@parity.io>

xermicus committed:
rename the trait method

Signed-off-by: xermicus <cyrill@parity.io>

xermicus commented on substrate/frame/revive/src/exec.rs:
```diff
@@ -454,6 +459,9 @@ pub trait Executable<T: Config>: Sized {
 	fn from_storage(code_hash: H256, gas_meter: &mut GasMeter<T>) -> Result<Self, DispatchError>;
 
+	/// Load the executable from EVM bytecode
+	fn from_init_code(code: Vec<u8>, owner: AccountIdOf<T>) -> Result<Self, DispatchError>;
```
I agree, renamed it.

xermicus commented on substrate/frame/revive/src/vm/evm.rs:
```diff
@@ -44,6 +64,11 @@ impl<T: Config> ContractBlob<T> {
 			return Err(<Error<T>>::BlobTooLarge.into());
 		}
 
+		// EIP-3541: Reject new contract code starting with the 0xEF byte
+		if code.first() == Some(&0xEF) {
+			return Err(<Error<T>>::CodeRejected.into());
+		}
```
`new_raw_checked` checks against the [EIP 7702 magic bytes](https://eips.ethereum.org/EIPS/eip-7702), which considers the first two bytes and is less strict.

xermicus committed:
Merge branch 'master' into sm/vm-instr

xermicus commented:
/cmd fmt

github-actions[bot] committed:
Update from github-actions[bot] running command 'fmt'

github-actions[bot] committed:
Update from github-actions[bot] running command 'prdoc --audience runtime_dev --bump major'

xermicus committed:
prdoc

Signed-off-by: xermicus <cyrill@parity.io>

dlebee commented:
Just curious isn't returning 2500000000000000_u64 to prevrando miss leading? 

Frontier returns 0, why 2500000000000000_u64 is better than 0? just curious?

Couldn't this lead to a contract believing that prevrando is present ?

/// Hard-coded value returned by the EVM `DIFFICULTY` opcode.
///
/// After Ethereum's Merge (Sept 2022), the `DIFFICULTY` opcode was redefined to return
/// `prevrandao`, a randomness value from the beacon chain. In Substrate pallet-revive
/// a fixed constant is returned instead for compatibility with contracts that still read this
/// opcode. The value is aligned with the difficulty hardcoded for PVM contracts.
pub(crate) const DIFFICULTY: u64 = 2500000000000000_u64;
